src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -351,19 +315,90 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
         {
             if (string.IsNullOrEmpty(value))
+            {
                 return value;
+            }
+            
+            bool foundSpaces = false;
+            int unsafeCount = 0;
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            // count them first
+            var encoding = Encoding.UTF8;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+
+                if (ch == ' ')
+                {
+                    foundSpaces = true;
+                }
+                else if (!IsUrlSafeChar(ch))
+                {
+                    if (ch < 128) // optimize for ASCII chars
+                    {
+                        unsafeCount++; // UTF-8 uses 1 byte for chars 0 to 127
+                    }
+                    else if (char.IsSurrogatePair(value, i)) // this method should handle the case where i is the last index
+                    {
+                        // surrogate code points are encoded differently separate than when alone, so we need to address that
+                        i++;
+                        unsafeCount += encoding.GetByteCount(&ch, 2);
+                    }
+                    else
+                    {
+                        unsafeCount += encoding.GetByteCount(&ch, 1);
+                    }
+                }
+            }
+
+            // nothing to expand?
+            if (!foundSpaces && unsafeCount == 0)
+            {
+                return value;
+            }
+            
+            byte[] bytes = encoding.GetBytes(value);
+            // expand not 'safe' characters into %XX, spaces to +s
+            byte[] encodedBytes = new byte[bytes.Length + unsafeCount * 2];
+            UrlEncode(bytes, 0, bytes.Length, encodedBytes);
+            return encoding.GetString(encodedBytes, 0, encodedBytes.Length);
         }
 
         public static byte[] UrlEncodeToBytes(byte[] value, int offset, int count)
         {
-            return UrlEncode(value, offset, count, true /* alwaysCreateNewReturnValue */);
+            if (!ValidateUrlEncodingParameters(value, offset, count))
+            {
+                return null;
+            }
+            
+            int unsafeCount = 0;
+
+            // count them first
+            for (int i = 0; i < count; i++)
+            {
+                char ch = (char)value[offset + i];
+
+                if (ch != ' ' && !IsUrlSafeChar(ch))
+                {
+                    unsafeCount++;
+                }
+            }
+            
+            if (unsafeCount == 0)
+            {
+                // nothing to encode, just make a copy of the array
+                byte[] result = new byte[count];
+                Array.Copy(value, offset, result, 0, count);
```

Comment:
Note: I just realized this change solves a pretty big issue where a copy of the entire array will be returned if no encoding is needed, since the old implementation uses `Array.Clone` to clone the entire array.